yjit/src/core.rs

```diff
@@ -2675,36 +2709,51 @@ pub fn gen_branch_stub_hit_trampoline(ocb: &mut OutlinedCb) -> CodePtr {
     // Since this trampoline is static, it allows code GC inside
     // branch_stub_hit() to free stubs without problems.
     asm_comment!(asm, "branch_stub_hit() trampoline");
-    let jump_addr = asm.ccall(
+    let stub_hit_ret = asm.ccall(
         branch_stub_hit as *mut u8,
         vec![
             C_ARG_OPNDS[0],
             C_ARG_OPNDS[1],
             EC,
         ]
     );
+    let jump_addr = asm.load(stub_hit_ret);
 
     // Restore caller-saved registers for stack temps
-    for &reg in caller_saved_temp_regs().iter().rev() {
-        asm.cpop_into(reg);
+    for &reg in caller_saved_temp_regs().rev() {
+        asm.cpop_into(Opnd::Reg(reg));
     }
 
     // Jump to the address returned by the branch_stub_hit() call
     asm.jmp_opnd(jump_addr);
 
+    // HACK: restoring the value of C_RET_REG clobbers the
+    // return value of branch_stub_hit we want to jump to,
+    // so we need a scratch regsiter. This unrechable `test`
+    // maintain liveness of C return reg so we get something
+    // else for the scratch register.
+    asm.test(stub_hit_ret, 0.into());
```

Comment:
This would waste fewer bytes on x86_64.

```suggestion
    asm.test(stub_hit_ret, stub_hit_ret);
```